zebra-network/src/peer/connection.rs

```diff
@@ -1122,6 +1226,62 @@ where
     }
 }
 
+impl<S, Tx> Connection<S, Tx> {
+    /// Update the connection state metrics for this connection,
+    /// using `extra_state_info` as additional state information.
+    fn update_state_metrics(&mut self, extra_state_info: impl Into<Option<String>>) {
+        let current_metrics_state = if let Some(extra_state_info) = extra_state_info.into() {
+            format!("{}::{}", self.state.command(), extra_state_info)
+        } else {
+            self.state.command()
+        };
+
+        if self.last_metrics_state.as_ref() == Some(&current_metrics_state) {
+            return;
+        }
+
+        self.erase_state_metrics();
+
+        // Set the new state
+        metrics::increment_gauge!(
+            "zebra.net.connection.state",
+            1.0,
+            "command" => current_metrics_state.clone(),
+            "addr" => self.metrics_label.clone(),
+        );
+
+        self.last_metrics_state = Some(current_metrics_state);
+    }
+
+    /// Erase the connection state metrics for this connection.
+    fn erase_state_metrics(&mut self) {
+        if self.last_metrics_state.is_some() {
+            metrics::gauge!(
+                "zebra.net.connection.state",
+                0.0,
+                "command" => self.last_metrics_state.take().expect("just checked for None"),
+                "addr" => self.metrics_label.clone(),
+            );
+        }
```

Comment:
Idea: a small suggestion to remove the need for the `expect`:

```suggestion
        if let Some(last_metrics_state) = self.last_metrics_state.take() {
            metrics::gauge!(
                "zebra.net.connection.state",
                0.0,
                "command" => last_metrics_state,
                "addr" => self.metrics_label.clone(),
            );
        }
```

Reply:
Thanks! I created that method in a refactor, but I missed this simplification.

Fixed in 799e378 in PR #3203.